### ISSUE TITLE
RSA key generated instead of ECDSA

### ISSUE THREAD

**Issue by majkrzak** [OpenSC/OpenSC#1411]
Hi, I've finnaly get mine epass 2003 with ECDSA support. I've to set it up using opensc 0.18.0, but somethings got wrong. It seems that I've got there half RSA half ECDSA key. @FeitianSmartcardReader nice :sweat_smile: 

```
~ $ pkcs15-init --create-pkcs15 --profile pkcs15+onepin
Using reader with a card: Feitian ePass2003 00 00
New User PIN.
Please enter User PIN: 
Please type again to verify: 
Unblock Code for New User PIN (Optional - press return for no PIN).
Please enter User unblocking PIN (PUK): 
Please type again to verify: 
```
```
~ $ pkcs15-init --generate-key EC:prime256v1 --key-usage sign --auth-id 1
Using reader with a card: Feitian ePass2003 00 00
User PIN [User PIN] required.
Please enter User PIN [User PIN]: 
```
```
~ $ pkcs15-tool --dump
Using reader with a card: Feitian ePass2003 00 00
PKCS#15 Card [OpenSC Card]:
        Version        : 0
        Serial number  : 237535F88004002F
        Manufacturer ID: EnterSafe
        Last update    : 20180630174130Z
        Flags          : EID compliant
PIN [User PIN]
        Object Flags   : [0x3], private, modifiable
        ID             : 01
        Flags          : [0x32], local, initialized, needs-padding
        Length         : min_len:4, max_len:16, stored_len:16
        Pad char       : 0x00
        Reference      : 1 (0x01)
        Type           : ascii-numeric
        Path           : 3f005015

Private EC Key [Private Key]
        Object Flags   : [0x3], private, modifiable
        Usage          : [0xC], sign, signRecover
        Access Flags   : [0x1D], sensitive, alwaysSensitive, neverExtract, local
        FieldLength    : 256
        Key ref        : 0 (0x0)
        Native         : yes
        Path           : 3f0050152900
        Auth ID        : 01
        ID             : cb9572b4c1c354b4e29130ffda0bac83342c9446
        MD:guid        : 275db3ea-66fe-d874-9a14-4f9a0954d893

Public RSA Key [Private Key]
        Object Flags   : [0x2], modifiable
        Usage          : [0xC0], verify, verifyRecover
        Access Flags   : [0x0]
        ModLength      : 255
        Key ref        : 0 (0x0)
        Native         : no
        Path           : 3f0050153000
        ID             : cb9572b4c1c354b4e29130ffda0bac83342c9446

```
```
~ $ pkcs15-tool --read-public-key cb9572b4c1c354b4e29130ffda0bac83342c9446 | openssl asn1parse 
Using reader with a card: Feitian ePass2003 00 00
    0:d=0  hl=2 l=  59 cons: SEQUENCE          
    2:d=1  hl=2 l=  13 cons: SEQUENCE          
    4:d=2  hl=2 l=   9 prim: OBJECT            :rsaEncryption
   15:d=2  hl=2 l=   0 prim: NULL              
   17:d=1  hl=2 l=  42 prim: BIT STRING ```

**Comment by frankmorgner** [comment]
@FeitianSmartcardReader could you please have a look into this?

**Comment by mouse07410** [comment]
According to the man page for `pkcs15-init`, it does not support EC keys (with the only exception being GOST that seems to be supported in the code):
```
. . . . .
pkcs15-init lets you generate a new key and store it on the card. You can do this using:

       pkcs15-init --generate-key " keyspec " --auth-id " nn

       where keyspec describes the algorithm and length of the key to be created, such as rsa/512. This will create a
       512 bit RSA key. Currently, only RSA key generation is supported. Note that cards usually support just a few
       different key lengths. Almost all cards will support 512 and 1024 bit keys, some will support 768 or 2048 as
       well.
```

Am I missing something here?

**Comment by FeitianSmartcardReader** [comment]
sorry for the late reply, allow me forward to our R&D check and back to you guys tomorrow, thanks

**Comment by jcpallise** [comment]
@FeitianSmartcardReader , no answser from your R&D about ECDSA support ?
mine seems not supporting this mechanism...

**Comment by majkrzak** [comment]
@jcpallise There is special version of epass2003 with should have ecc onboard, but it is sold from under the table and you have to ask about it.

**Comment by dengert** [comment]
The above `pkcs15-tool -dump` lists two private keys, an EC key and a RSA key **but both have the same ID**: `cb9572b4c1c354b4e29130ffda0bac83342c9446` That look wrong. 

So the `pkcs15-tool --read-public-key cb9572b4c1c354b4e29130ffda0bac83342c9446` may be reading the RSA key. 

A opensc-debug.log of the `pkcs15-init --generate-key EC:prime256v1 --key-usage sign --auth-id 1` 
command would show if a EC or RSA key  is generated, based on the returned data from the card. 
It might show how the ID is generated too. The EC would be much shorter. 
then try listing teh public keys too. 

Do you need to add the `--id <arg>` parameter to the  `pkcs15-init`  command?

**Comment by FeitianSmartcardReader** [comment]
We support ECDSA with non-FIPS mode to support ECC. And the certificate you showed us is new version token, you can contact Feitian to get ECDSA token with FIPS mode.

**Comment by FeitianSmartcardReader** [comment]
@majkrzak issue solved, already submit pull request
https://github.com/OpenSC/OpenSC/pull/1488/commits/6d50184724f4508dfec468851ae7ea08ca29839d

**Comment by majkrzak** [comment]
Thank you, now we have to wait for Frank :P

czw., 27 wrz 2018, 12:50 użytkownik Feitian Technologies <
notifications@github.com> napisał:

**Comment by FeitianSmartcardReader** [comment]
@majkrzak Merged already, take a look and change the bug status, thanks

**Comment by Jakuje** [comment]
I think this was fixed in the previous release. If not, please, open a new issue.
